### PR TITLE
[4] Fix undefined classes in params

### PR DESCRIPTION
**Pull request by PhilETaylor**
Code review

### CODE REVIEW

**Comment by PhilETaylor** [comment]
Drone failure unrelated.

File: components/com_contenthistory/src/Controller/DisplayController.php [hunk 1]
```diff
@@ -28,7 +29,7 @@ class DisplayController extends BaseController
 	 * Recognized key values include 'name', 'default_task', 'model_path', and
 	 * 'view_path' (this list is not meant to be comprehensive).
 	 * @param   MVCFactoryInterface  $factory  The factory.
-	 * @param   CMSApplication       $app      The JApplication for the dispatcher
+	 * @param   CMSApplication|null  $app      The JApplication for the dispatcher
 	 * @param   \JInput              $input    Input
```
Comment (alikon) [inline components/com_contenthistory/src/Controller/DisplayController.php]: shouldn't be done the same for Input ?

Comment (PhilETaylor) [reply components/com_contenthistory/src/Controller/DisplayController.php]: and a million other places in joomla - yes.

File: components/com_contenthistory/src/Controller/DisplayController.php [hunk 2]
```diff
@@ -27,9 +28,9 @@ class DisplayController extends BaseController
 	 * @param   array                $config   An optional associative array of configuration settings.
 	 * Recognized key values include 'name', 'default_task', 'model_path', and
 	 * 'view_path' (this list is not meant to be comprehensive).
-	 * @param   MVCFactoryInterface  $factory  The factory.
-	 * @param   CMSApplication       $app      The JApplication for the dispatcher
-	 * @param   \JInput              $input    Input
+	 * @param   MVCFactoryInterface|null  $factory  The factory.
+	 * @param   CMSApplication|null       $app      The JApplication for the dispatcher
+	 * @param   \JInput|null              $input    Input
```
Comment (alikon) [inline components/com_contenthistory/src/Controller/DisplayController.php]:
```suggestion
	 * @param   Input|null              $input    Input
```

File: components/com_contenthistory/src/Controller/DisplayController.php [hunk 3]
```diff
@@ -11,6 +11,7 @@
 
 \defined('_JEXEC') or die;
 
+use Joomla\CMS\Application\CMSApplication;
```
Comment (alikon) [inline components/com_contenthistory/src/Controller/DisplayController.php:14]:
```suggestion
use Joomla\CMS\Application\CMSApplication;
use Joomla\CMS\Input\Input;
```

File: components/com_contenthistory/src/Controller/DisplayController.php (hunk 2 above)
Comment (PhilETaylor) [reply components/com_contenthistory/src/Controller/DisplayController.php]: That can only be done if the class is imported too

**Comment by PhilETaylor** [comment]
This PR was originally opened to just fix the undefined class in the param that phpStorm moans about. Now, with the proposed changes, it does more than that (as `JInput` was not undefined :))

**Comment by alikon** [comment]
i moaning more than phpStorm :smiley:

**Comment by alikon** [comment]
I have tested this item :white_check_mark: successfully on 3525fe8044eb04af6443c404069230f96269dc1a<hr /><sub>This comment was created with the <a href="https://github.com/joomla/jissues">J!Tracker Application</a> at <a href="https://issues.joomla.org/tracker/joomla-cms/35189">issues.joomla.org/tracker/joomla-cms/35189</a>.</sub>

**Comment by richard67** [comment]
@PhilETaylor Drone reports code style errors: https://ci.joomla.org/joomla/joomla-cms/46918/1/6

**Comment by PhilETaylor** [comment]
fixed

**Comment by richard67** [comment]
I've restored the previous test result since the change after that was just in a comment.

**Comment by richard67** [comment]
I have tested this item :white_check_mark: successfully on 862fdff175e6fd064e05ec932701027eac12ba4c<br /><br />Code review.<hr /><sub>This comment was created with the <a href="https://github.com/joomla/jissues">J!Tracker Application</a> at <a href="https://issues.joomla.org/tracker/joomla-cms/35189">issues.joomla.org/tracker/joomla-cms/35189</a>.</sub>

**Comment by richard67** [comment]
RTC<hr /><sub>This comment was created with the <a href="https://github.com/joomla/jissues">J!Tracker Application</a> at <a href="https://issues.joomla.org/tracker/joomla-cms/35189">issues.joomla.org/tracker/joomla-cms/35189</a>.</sub>

File: components/com_contenthistory/src/Controller/DisplayController.php [hunk 4]
```diff
@@ -22,14 +24,10 @@
...
-	 * @param   array                $config   An optional associative array of configuration settings.
-	 * Recognized key values include 'name', 'default_task', 'model_path', and
-	 * 'view_path' (this list is not meant to be comprehensive).
-	 * @param   MVCFactoryInterface  $factory  The factory.
-	 * @param   CMSApplication       $app      The JApplication for the dispatcher
-	 * @param   \JInput              $input    Input
+	 * @param   array                     $config   An optional associative array of configuration settings.
+	 * @param   MVCFactoryInterface|null  $factory  The factory.
+	 * @param   CMSApplication|null       $app      The JApplication for the dispatcher
+	 * @param   Input|null                $input    Input
```
Comment (wilsonge) [inline components/com_contenthistory/src/Controller/DisplayController.php]:
```suggestion
	 * @param   ?Input                $input    Input
```

**Comment by wilsonge** [comment]
Thanks!

File: components/com_contenthistory/src/Controller/DisplayController.php [hunk 5]
```diff
@@ -27,7 +27,7 @@ class DisplayController extends BaseController
 	 * @param   array                     $config   An optional associative array of configuration settings.
 	 * @param   MVCFactoryInterface|null  $factory  The factory.
 	 * @param   CMSApplication|null       $app      The JApplication for the dispatcher
-	 * @param   Input|null                $input    Input
+	 * @param   ?Input                $input    Input
```
Comment (richard67) [inline components/com_contenthistory/src/Controller/DisplayController.php:30]: @wilsonge Please fix code style :-(

Comment (PhilETaylor) [reply components/com_contenthistory/src/Controller/DisplayController.php:30]: Absolutely hate `?Type` is it even a valid value in a type in a docblock?

Comment (PhilETaylor) [reply components/com_contenthistory/src/Controller/DisplayController.php:30]: apparently its valid :-( but still changing it at the last min when this had already had 2 testers and fixed cs issues previously was a bit harsh... This syntax is only used in 7 other places in Joomla and libs.

Comment (richard67) [reply components/com_contenthistory/src/Controller/DisplayController.php:30]: @wilsonge What I don't understand is why you haven't changed it for the other nullable parameters, too.

**Pull request merged**